command/013_config_upgrade.go

```diff
@@ -71,58 +75,386 @@ func (c *ZeroThirteenUpgradeCommand) Run(args []string) int {
 		return 1
 	}
 
-	// Early-load the config so that we can check provider dependencies
-	earlyConfig, earlyConfDiags := c.loadConfigEarly(dir)
-	if earlyConfDiags.HasErrors() {
+	// Set up the config loader and find all the config files
+	loader, err := c.initConfigLoader()
+	if err != nil {
+		diags = diags.Append(err)
+		c.showDiagnostics(diags)
+		return 1
+	}
+	parser := loader.Parser()
+	primary, overrides, hclDiags := parser.ConfigDirFiles(dir)
+	diags = diags.Append(hclDiags)
+	if diags.HasErrors() {
 		c.Ui.Error(strings.TrimSpace("Failed to load configuration"))
-		diags = diags.Append(earlyConfDiags)
 		c.showDiagnostics(diags)
 		return 1
 	}
 
-	{
-		// Before we go further, we'll check to make sure none of the modules
-		// in the configuration declare that they don't support this Terraform
-		// version, so we can produce a version-related error message rather
-		// than potentially-confusing downstream errors.
-		versionDiags := initwd.CheckCoreVersionRequirements(earlyConfig)
```

Comment:
I'm curious why this was removed (unless I've completely missed it!): we don't want to upgrade a configuration if it is still hard-coded to an earlier version of terraform.

Reply:
Whoops, I didn't explain what was going on here at all.

I removed this because we no longer have an implementation of `CheckCoreVersionRequirements` that we can use. We have neither a `configs.Config` nor an `earlyconfig.Config`, because we're loading and parsing each file at a lower level.

Now that you point out the problem here, it's clear to me that it's worth going back to address it. I'm not immediately sure how to do so, but I think it will overlap with @apparentlymart's suggestion to update the `required_version` string in `versions.tf` if it's present. I plan to revisit that in a follow-up PR.

Reply:
I believe we had this check in `0.12upgrade` as part of our heuristics to help us give better feedback if the user tries to re-run the upgrade process on an already-upgraded configuration.

I'm not sure we have exactly the same need here, because I _think_ the `0.13upgrade` process is idempotent, right? Specifically, I'd expect re-running it to find that there is only one `required_providers` block and that everything inside it already has `source` set and therefore exit without changing anything.

It could be valuable to check specifically that the configuration isn't requiring a _later_ version of Terraform, like `>= 0.14`, in case the syntax changes in a way that would make this upgrade process produce a confusing error. But it also sounds like it might be complex to do so within the strategy we're employing here, and so I think it's justified to leave this out under the assumption that running an old upgrade tool against a newer config is not something a _lot_ of users are likely to encounter.

Reply:
Yes, the upgrader is idempotent (or at least, if it isn't, that's a bug).

After thinking about this a bit more, I don't think checking the version constraint here is necessary. If a user runs the upgrader on a config which has `required_version = "
~> 0.12"` or similar, I think it's fair to assume that their intention is still to upgrade the configuration. Blocking them until they fix the version pinning seems overly pedantic.

I've made a note to revisit this when we consider adding a new required version constraint to the `versions.tf` file, and I'm happy to discuss it further. Resolving for now!